### PR TITLE
Add additional capability (filter by decay R/Z) to MCMultiParticleFilter

### PR DESCRIPTION
**Pull request by kfjack**
#### PR description:

For the B->mumu analysis, we would like to generate a special MC which has a full decay chain of B(s,d) -> hh(=k,pi) -> mumu, with the hadrons decaying to the muon within the detector volume. We have identified such a MC can be produced if an additional capability (filtering according to the decay region, the same functionality which has been already implemented in the MCSmartSingleParticleFilter) can be added to the MCMultiParticleFilter. Hence this PR is proposed, with no change to the original filter, but just adding new capabilities.  

A small typo (bug, but no real effect) for loading motherID_  in the original filter has been fixed as well.

#### PR validation:

Managed to filter the B(s,d) -> hh -> mumu events by adding such a filter (requiring two muon decaying from K/pi, and within the detector volume) to the process:
mugenfilter = cms.EDFilter(
    "MCMultiParticleFilter",
    ParticleID = cms.vint32(-13, 13, -13, 13),
    MotherID = cms.untracked.vint32(321, -321, 211, -211),
    MaxDecayRadius = cms.untracked.vdouble(2000.0, 2000.0, 2000.0, 2000.0),
    MaxDecayZ = cms.untracked.vdouble(4000.0, 4000.0, 4000.0, 4000.0),
    MinDecayZ = cms.untracked.vdouble(-4000.0, -4000.0, -4000.0, -4000.0),
    EtaMax = cms.vdouble(2.5, 2.5, 2.5, 2.5),
    PtMin = cms.vdouble(3.5, 3.5, 3.5, 3.5),
    Status = cms.vint32(1,1,1,1),
    NumRequired = cms.int32(2),
    AcceptMore = cms.bool(True)
)

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Not a backport PR.

### CODE REVIEW

File: GeneratorInterface/GenFilters/plugins/MCMultiParticleFilter.cc [hunk 1]
```diff
@@ -96,6 +121,40 @@ MCMultiParticleFilter::MCMultiParticleFilter(const edm::ParameterSet& iConfig)
...
+    decayZMin = decayZmin2;
+  }
+  // if decayZMax size smaller than particleID , fill up further with defaults
+  if (particleID_.size() > decayZMax.size()) {
+    std::vector<double> decayZmax2;
+    for (unsigned int i = 0; i < particleID_.size(); i++) {
+      decayZmax2.push_back(1.e5);
+    }
+    decayZMax = decayZmax2;
+  }
```
Comment (agrohsje) [inline GeneratorInterface/GenFilters/plugins/MCMultiParticleFilter.cc:157]: Doesn't that mean that you will also overwrite already existing values and not only fill up the vector.

Comment (agrohsje) [reply GeneratorInterface/GenFilters/plugins/MCMultiParticleFilter.cc:157]: But wouldn't it be more suitable to just add the defaults and not replace all? This is also what the description says. The code just doesn't do it.

Comment (kfjack) [reply GeneratorInterface/GenFilters/plugins/MCMultiParticleFilter.cc:157]: Many thanks. In fact I fully agree that it is more suitable to just add the defaults on top of the existing vector –– if different implementations between two filters are allowed, I'll change this (or I should also purpose to change MCSmartSingleParticleFilter at the same time?). Please let me know.

**Comment by agrohsje** [comment]
I think it would be better to adjust both. Maybe @qliphy can comment on his preferred choice.

**Comment by qliphy** [comment]
> I think it would be better to adjust both. Maybe @qliphy can comment on his preferred choice.

I agree adjusting both to add the defaults on top of the existing vector is indeed better.

**Comment by kfjack** [comment]
Now the PR is updated accordingly: fill the vectors with defaults if necessary for both filters. Many thanks.

**Comment by agrohsje** [comment]
please test

**Comment by kfjack** [comment]
DAS error is not an issue related to the code itself, right? Can we test it again? Many thanks!

**Comment by qliphy** [comment]
> DAS error is not an issue related to the code itself, right? Can we test it again? Many thanks!

Yes, the DAS error is due to PPS workflows discussed in: https://github.com/cms-sw/cmssw/issues/33010
You can ignore the error. @cms-sw/generators-l2

**Comment by agrohsje** [comment]
+generators

**Comment by qliphy** [comment]
+1

**Comment by qliphy** [comment]
merge

**Pull request merged**